src/main/java/bio/terra/service/resourcemanagement/google/GoogleResourceService.java

```diff
@@ -68,44 +77,148 @@ public GoogleResourceService(
+        this.configService = configService;
     }
 
+    /**
+     * Fetch an existing bucket_resource metadata row.
+     * Note this method does not check for the existence of the underlying cloud resource.
+     * @param bucketResourceId
+     * @returna a reference to the bucket as a POJO GoogleBucketResource
+     */
     public GoogleBucketResource getBucketResourceById(UUID bucketResourceId) {
         return resourceDao.retrieveBucketById(bucketResourceId);
     }
 
-    public GoogleBucketResource getOrCreateBucket(GoogleBucketRequest bucketRequest) {
-        // see if there is a bucket resource that matches the project and bucket name from the request
-        GoogleProjectResource projectResource = bucketRequest.getGoogleProjectResource();
-        try {
-            Optional<GoogleBucketResource> possibleMatch = resourceDao.retrieveBucketsByProjectResource(projectResource)
-                .stream()
-                .filter(bucketResource -> bucketResource.getName().equals(bucketRequest.getBucketName()))
-                .findFirst();
-            if (possibleMatch.isPresent()) {
-                return possibleMatch.get();
+    /**
+     * Fetch/create a bucket cloud resource and the associated metadata in the bucket_resource table.
+     * The flightId is used to lock the bucket metadata during possible creation.
+     * Below is an outline of the logic flow.
+     *   1. Try to insert a new bucket_resource row and lock it
+     *      If the insert failed, lookup the existing row
+     *   2. Check who holds the lock on the bucket_resource row
+     *      a. GET case = row is unlocked. bucket has already been created
+     *         Check that the bucket cloud resource exists, throw an exception if not.
+     *      b. CREATE case = row is locked by this flight. need to create the bucket here
+     *         Check if the bucket cloud resource exists.
+     *           i. If not, create it
+     *           ii. If it does exist and this is a testing environment, just use it
+     *           iii. If it does exist and this is a recovery flight, just use it
+     *           iv. Otherwise, the state of the cloud is out of sync with the metadat
+     *               Delete anything we've created and throw an exception
+     *      c. LOCKED case = the row is locked by another flight. throw a lock exception so the flight knows to retry
+     * @param bucketRequest
+     * @param flightId
+     * @return a reference to the bucket as a POJO GoogleBucketResource
+     */
+    public GoogleBucketResource getOrCreateBucket(GoogleBucketRequest bucketRequest, String flightId) {
+        // insert a new bucket_resource row and lock it
+        GoogleBucketResource googleBucketResource = resourceDao.createAndLockBucket(bucketRequest, flightId);
+
+        boolean metadataCreationFailed = (googleBucketResource == null);
+        logger.info("metadataCreationFailed = " + metadataCreationFailed);
```

Comment:
Changed some of my additions to debug logging. Kept just the ones for the fault since they should only be triggered when debugging anyway, and the log of the application property.